In [1]:
!nvidia-smi --query-gpu=name,memory.total,memory.free --format=csv


name, memory.total [MiB], memory.free [MiB]
Tesla T4, 15360 MiB, 14913 MiB


In [3]:
from google.colab import files
import os, shutil

print("Re-uploading kaggle.json (find it in your Downloads folder)...")
uploaded = files.upload()

os.makedirs('/root/.kaggle', exist_ok=True)
shutil.move('kaggle.json', '/root/.kaggle/kaggle.json')
os.chmod('/root/.kaggle/kaggle.json', 0o600)
print("kaggle.json configured.")

Re-uploading kaggle.json (find it in your Downloads folder)...


Saving kaggle.json to kaggle.json
kaggle.json configured.


In [4]:
"""
SESSION BOOTSTRAP — Phase 4 (Stage 2)
Restores: Drive, repo, dataset, Python imports, wandb auth.
"""
import os, sys, time

print("=" * 60)
print("Step 1/4: Mount Drive")
print("=" * 60)
DRIVE_ROOT = '/content/drive/MyDrive'
if not os.path.exists(DRIVE_ROOT):
    from google.colab import drive
    drive.mount('/content/drive')
print("Drive mounted.\n")

os.makedirs('/content/drive/MyDrive/deepfake_audio/checkpoints', exist_ok=True)
os.makedirs('/content/drive/MyDrive/deepfake_audio/logs', exist_ok=True)

print("=" * 60)
print("Step 2/4: Clone/update repo")
print("=" * 60)
REPO_DIR = '/content/deepfake-audio-detection'
if not os.path.exists(REPO_DIR):
    !git clone https://github.com/Saracasm/deepfake-audio-detection.git {REPO_DIR}
else:
    !cd {REPO_DIR} && git pull --quiet
print(f"Repo at: {REPO_DIR}\n")

print("=" * 60)
print("Step 3/4: Re-download dataset (~3-5 min)")
print("=" * 60)
LOCAL_LA = '/content/kaggle_download/LA'

if os.path.exists(LOCAL_LA):
    print("Dataset already present.")
else:
    if not os.path.exists('/root/.kaggle/kaggle.json'):
        print("ERROR: kaggle.json not configured.")
        print("Run the kaggle.json upload cell BEFORE this bootstrap.")
        raise SystemExit("Need kaggle credentials")

    !pip install -q kaggle
    os.makedirs('/content/kaggle_download', exist_ok=True)
    start = time.time()
    !kaggle datasets download -d anishsarkar22/asvpoof-2019-dataset-la \
        -p /content/kaggle_download --unzip --force --quiet
    print(f"Downloaded in {(time.time()-start)/60:.1f} minutes.")

print(f"Dataset at: {LOCAL_LA}\n")

print("=" * 30)
print("Step 4/4: Set up Python imports + wandb")
print("=" * 30)
sys.path.insert(0, REPO_DIR)
LA_ROOT = LOCAL_LA

# Wandb key from Colab Secrets (so we don't hit interactive prompt)
try:
    from google.colab import userdata
    os.environ['WANDB_API_KEY'] = userdata.get('WANDB_API_KEY')
    print("Wandb API key loaded from Colab Secrets.")
except Exception as e:
    print(f"WANDB_API_KEY not loaded: {e}")

print(f"\nLA_ROOT = {LA_ROOT}")
print(f"REPO_DIR = {REPO_DIR}")
print("\nBootstrap complete. Ready to work.")

Step 1/4: Mount Drive
Drive mounted.

Step 2/4: Clone/update repo
Repo at: /content/deepfake-audio-detection

Step 3/4: Re-download dataset (~3-5 min)
Dataset URL: https://www.kaggle.com/datasets/anishsarkar22/asvpoof-2019-dataset-la
License(s): ODC Attribution License (ODC-By)
Downloaded in 2.6 minutes.
Dataset at: /content/kaggle_download/LA

Step 4/4: Set up Python imports + wandb
Wandb API key loaded from Colab Secrets.

LA_ROOT = /content/kaggle_download/LA
REPO_DIR = /content/deepfake-audio-detection

Bootstrap complete. Ready to work.


In [5]:
import sys, importlib, torch

# Make sure imports come from latest repo files
sys.path.insert(0, '/content/deepfake-audio-detection')
for mod in ['src.data.protocols', 'src.data.preprocessing', 'src.data.dataset',
            'src.models.wav2vec_classifier', 'src.evaluation.metrics',
            'src.training.trainer']:
    if mod in sys.modules:
        importlib.reload(sys.modules[mod])

from src.data.protocols import parse_all_partitions, class_counts
from src.data.dataset import ASVspoofDataset
from src.models.wav2vec_classifier import Wav2VecClassifier

# Re-parse protocols (fast)
LA_ROOT = '/content/kaggle_download/LA'
splits = parse_all_partitions(LA_ROOT)
print("Protocols re-parsed:")
for name, utts in splits.items():
    counts = class_counts(utts)
    print(f"  {name}: {len(utts):,} ({counts['bonafide']:,} bonafide, {counts['spoof']:,} spoof)")

# Verify Stage 1 checkpoint
device = 'cuda' if torch.cuda.is_available() else 'cpu'
print(f"\nDevice: {device}\n")

CKPT_PATH = '/content/drive/MyDrive/deepfake_audio/checkpoints/stage1_best.pt'
import os
print(f"Stage 1 checkpoint exists: {os.path.exists(CKPT_PATH)}")
print(f"Size: {os.path.getsize(CKPT_PATH) / 1e6:.1f} MB")

ckpt = torch.load(CKPT_PATH, map_location='cpu', weights_only=False)
print(f"\nCheckpoint contents:")
print(f"  Epoch saved:  {ckpt['epoch']}")
print(f"  Best dev EER: {ckpt['best_eer']*100:.2f}%")
print(f"  Param tensors in state_dict: {len(ckpt['model_state_dict'])}")

# Build a fresh model and load the Stage 1 weights into it
print("\nLoading Stage 1 weights into fresh model...")
model = Wav2VecClassifier(
    backbone_name="facebook/wav2vec2-base",
    num_classes=2,
    freeze_backbone=True,
)
missing, unexpected = model.load_state_dict(ckpt['model_state_dict'], strict=False)
print(f"  Missing keys: {len(missing)}")
print(f"  Unexpected keys: {len(unexpected)}")

if len(missing) == 0 and len(unexpected) == 0:
    print("\n✓ Stage 1 weights loaded perfectly. Ready for Stage 2.")
else:
    print("\nWARNING: state_dict mismatch — investigate before proceeding.")
    if missing:
        print(f"  Missing: {missing[:5]}")
    if unexpected:
        print(f"  Unexpected: {unexpected[:5]}")

model = model.to(device)
print(f"\nModel on {device}, ready.")


Protocols re-parsed:
  train: 25,380 (2,580 bonafide, 22,800 spoof)
  dev: 24,844 (2,548 bonafide, 22,296 spoof)
  eval: 71,237 (7,355 bonafide, 63,882 spoof)

Device: cuda

Stage 1 checkpoint exists: True
Size: 377.6 MB

Checkpoint contents:
  Epoch saved:  5
  Best dev EER: 10.09%
  Param tensors in state_dict: 213

Loading Stage 1 weights into fresh model...


/usr/local/lib/python3.12/dist-packages/huggingface_hub/utils/_auth.py:93: UserWarning: 
The secret `HF_TOKEN` does not exist in your Colab secrets.
To authenticate with the Hugging Face Hub, create a token in your settings tab (https://huggingface.co/settings/tokens), set it as secret in your Google Colab and restart your session.
You will be able to reuse this secret in all of your notebooks.
Please note that authentication is recommended but still optional to access public models or datasets.
  warnings.warn(


Wav2Vec2Model LOAD REPORT from: facebook/wav2vec2-base
Key                          | Status     |  | 
-----------------------------+------------+--+-
quantizer.codevectors        | UNEXPECTED |  | 
quantizer.weight_proj.weight | UNEXPECTED |  | 
project_q.weight             | UNEXPECTED |  | 
project_q.bias               | UNEXPECTED |  | 
project_hid.bias             | UNEXPECTED |  | 
project_hid.weight           | UNEXPECTED |  | 
quantizer.weight_proj.bias   | UNEXPECTED |  | 

Notes:
- UNEXPECTED	:can be ignored when loading from different task/architecture; not ok if you expect identical arch.


  Missing keys: 0
  Unexpected keys: 0

✓ Stage 1 weights loaded perfectly. Ready for Stage 2.

Model on cuda, ready.


In [6]:
# Show parameter counts BEFORE unfreezing (Stage 1 state)
total_before = sum(p.numel() for p in model.parameters())
trainable_before = sum(p.numel() for p in model.parameters() if p.requires_grad)
print(f"Before unfreezing (Stage 1 state):")
print(f"  Total:     {total_before:>12,}")
print(f"  Trainable: {trainable_before:>12,}  ({100*trainable_before/total_before:.4f}%)")
print(f"  → Just the linear head (1,538 params)")

# Apply Stage 2 unfreeze
print(f"\nApplying unfreeze_top_n_layers(2)...")
model.unfreeze_top_n_layers(2)

# Recount
total_after = sum(p.numel() for p in model.parameters())
trainable_after = sum(p.numel() for p in model.parameters() if p.requires_grad)
print(f"\nAfter unfreezing top 2 transformer layers + final LayerNorm:")
print(f"  Total:     {total_after:>12,}")
print(f"  Trainable: {trainable_after:>12,}  ({100*trainable_after/total_after:.4f}%)")

# Sanity check: each transformer layer in Wav2Vec2-Base has ~7.1M params
# 2 layers + LayerNorm + classifier head should be roughly 14-15M trainable
expected_min = 14_000_000
expected_max = 16_000_000
print(f"\nExpected range: {expected_min:,} - {expected_max:,}")
print(f"Match: {expected_min <= trainable_after <= expected_max}")

# Verify which specific layers are now trainable
print(f"\nDetailed breakdown of TRAINABLE parameters:")
for name, p in model.named_parameters():
    if p.requires_grad:
        # Don't print every weight - just show layer-level summary
        # Show parent module path (e.g. 'backbone.encoder.layers.10' instead of full)
        parts = name.split('.')
        if 'layers' in parts:
            layer_idx = parts.index('layers') + 1
            parent = '.'.join(parts[:layer_idx + 1])  # e.g. backbone.encoder.layers.10
        else:
            parent = '.'.join(parts[:-1])
        # Show one line per parameter (gets verbose but instructive)
        print(f"  {name:<60} ({p.numel():,} params)")

Before unfreezing (Stage 1 state):
  Total:       94,373,250
  Trainable:        1,538  (0.0016%)
  → Just the linear head (1,538 params)

Applying unfreeze_top_n_layers(2)...

After unfreezing top 2 transformer layers + final LayerNorm:
  Total:       94,373,250
  Trainable:   14,178,818  (15.0242%)

Expected range: 14,000,000 - 16,000,000
Match: True

Detailed breakdown of TRAINABLE parameters:
  backbone.encoder.layer_norm.weight                           (768 params)
  backbone.encoder.layer_norm.bias                             (768 params)
  backbone.encoder.layers.10.attention.k_proj.weight           (589,824 params)
  backbone.encoder.layers.10.attention.k_proj.bias             (768 params)
  backbone.encoder.layers.10.attention.v_proj.weight           (589,824 params)
  backbone.encoder.layers.10.attention.v_proj.bias             (768 params)
  backbone.encoder.layers.10.attention.q_proj.weight           (589,824 params)
  backbone.encoder.layers.10.attention.q_proj.bias      

What's now trainable (Stage 2)14,178,818 parameters across:
2 transformer encoder layers (layers 10 and 11 of 12) — the top two. Each has:

Multi-head attention (Q, K, V, output projections): ~2.36M params
Feed-forward network (768 → 3072 → 768): ~4.72M params
Layer norms: ~3K params
Total per layer: ~7.08M params



Final encoder layer norm: 1,536 params

Classifier head: 1,538 params (unchanged from Stage 1 — your trained head from yesterday is the starting point, not random)
Frozen (80M params): The bottom 10 transformer layers and the convolutional feature extractor that turns raw waveform into framewise features. These were learned from 60,000 hours of unlabeled speech and are doing low-level acoustic processing — we don't want to disturb them.Why this architecture choice matters for your reportTwo-stage fine-tuning is a real strategy from the speech-recognition literature. The reasoning:
Lower layers learn general features (phoneme-like structure, spectral patterns). These transfer well across tasks.
Higher layers learn task-specific abstractions. When you switch from "predict masked audio frames" (the pretraining task) to "detect synthetic speech" (your task), it's the high-level representations that need to adapt most.
Freezing 80M params and tuning 14M is a regularization choice. With only 25k training utterances, fine-tuning all 94M would risk catastrophic overfitting. By freezing most of the model, you constrain capacity to what your data can support.

Mixed precision (fp16) — speeds up training ~30-40% on T4 by using half-precision floats where safe
Warmup + linear decay LR scheduler — fine-tuning at very low LRs benefits from gradual warmup to avoid destabilizing the pretrained weights in the first few steps
Gradient accumulation — optional, lets us simulate larger batch sizes if needed

In [7]:
TRAINER_PY = '''"""
Training loop for Wav2Vec-based deepfake audio detection.

Supports both Stage 1 (frozen backbone, simple) and Stage 2 (fine-tuning,
with mixed precision + warmup scheduler).
"""

import os
import time
from dataclasses import dataclass, field
from typing import Optional, Callable, List

import numpy as np
import torch
import torch.nn as nn
from torch.utils.data import DataLoader
from tqdm import tqdm

from src.evaluation.metrics import (
    compute_eer,
    compute_auc,
    aggregate_window_scores_to_utterance,
)


@dataclass
class TrainConfig:
    """Hyperparameters for one training stage."""
    learning_rate: float = 1e-3
    batch_size: int = 32
    epochs: int = 5
    weight_decay: float = 0.01
    grad_clip: float = 1.0
    early_stopping_patience: int = 3
    checkpoint_dir: str = "/content/drive/MyDrive/deepfake_audio/checkpoints"
    checkpoint_name: str = "stage1_best.pt"
    log_every_n_steps: int = 20
    use_wandb: bool = True
    wandb_project: str = "deepfake-audio-detection"
    wandb_run_name: Optional[str] = None
    class_weights: Optional[List[float]] = None
    # Stage 2 additions
    use_mixed_precision: bool = False
    warmup_ratio: float = 0.0  # fraction of total steps used for LR warmup
    use_lr_scheduler: bool = False  # set True with warmup_ratio > 0


def make_loss_fn(class_weights: Optional[List[float]], device: str) -> Callable:
    if class_weights is not None:
        weights = torch.tensor(class_weights, dtype=torch.float32, device=device)
        return nn.CrossEntropyLoss(weight=weights)
    return nn.CrossEntropyLoss()


def make_lr_scheduler(optimizer, total_steps: int, warmup_ratio: float):
    """Linear warmup followed by linear decay to zero."""
    warmup_steps = int(total_steps * warmup_ratio)
    def lr_lambda(step):
        if step < warmup_steps:
            return step / max(1, warmup_steps)
        progress = (step - warmup_steps) / max(1, total_steps - warmup_steps)
        return max(0.0, 1.0 - progress)
    return torch.optim.lr_scheduler.LambdaLR(optimizer, lr_lambda)


@torch.no_grad()
def evaluate(
    model: nn.Module,
    dev_loader: DataLoader,
    device: str,
    desc: str = "Eval",
    use_mixed_precision: bool = False,
) -> dict:
    """Run inference over the dev set and compute per-utterance metrics."""
    model.eval()
    all_window_scores, all_window_labels, all_window_utts = [], [], []
    total_loss, n_batches = 0.0, 0
    loss_fn = nn.CrossEntropyLoss()

    autocast_ctx = torch.amp.autocast(device_type="cuda", enabled=use_mixed_precision)

    for waveforms, labels, utt_ids in tqdm(dev_loader, desc=desc, leave=False):
        waveforms = waveforms.to(device, non_blocking=True)
        labels_gpu = labels.to(device, non_blocking=True)

        with autocast_ctx:
            logits = model(waveforms)
            loss = loss_fn(logits, labels_gpu)
        total_loss += loss.item()
        n_batches += 1

        probs = torch.softmax(logits.float(), dim=-1)
        spoof_probs = probs[:, 1].detach().cpu().numpy()

        all_window_scores.extend(spoof_probs.tolist())
        all_window_labels.extend(labels.tolist())
        all_window_utts.extend(list(utt_ids))

    utt_scores, utt_ids_sorted = aggregate_window_scores_to_utterance(
        np.array(all_window_scores), all_window_utts, method="mean",
    )
    label_map = {}
    for s, l, u in zip(all_window_scores, all_window_labels, all_window_utts):
        if u not in label_map:
            label_map[u] = l
    utt_labels = np.array([label_map[u] for u in utt_ids_sorted])

    eer, threshold = compute_eer(utt_scores, utt_labels)
    auc = compute_auc(utt_scores, utt_labels)
    preds = (utt_scores > threshold).astype(int)
    accuracy = float((preds == utt_labels).mean())

    return {
        "eer": eer, "auc": auc, "accuracy": accuracy,
        "threshold": float(threshold),
        "loss": total_loss / max(n_batches, 1),
        "n_utterances": len(utt_ids_sorted),
    }


def train(
    model: nn.Module,
    train_loader: DataLoader,
    dev_loader: DataLoader,
    config: TrainConfig,
    device: str = "cuda",
) -> dict:
    """Train the model for `config.epochs` epochs, evaluating each epoch."""
    wandb = None
    if config.use_wandb:
        import wandb as _wandb
        run = _wandb.init(
            project=config.wandb_project,
            name=config.wandb_run_name,
            config={
                "learning_rate": config.learning_rate,
                "batch_size": config.batch_size,
                "epochs": config.epochs,
                "weight_decay": config.weight_decay,
                "use_mixed_precision": config.use_mixed_precision,
                "warmup_ratio": config.warmup_ratio,
                "trainable_params": sum(p.numel() for p in model.parameters() if p.requires_grad),
            },
            settings=_wandb.Settings(init_timeout=180),
        )
        wandb = _wandb

    trainable = [p for p in model.parameters() if p.requires_grad]
    optimizer = torch.optim.AdamW(
        trainable,
        lr=config.learning_rate,
        weight_decay=config.weight_decay,
    )

    # Optional LR scheduler with warmup
    scheduler = None
    if config.use_lr_scheduler and config.warmup_ratio > 0:
        total_steps = len(train_loader) * config.epochs
        scheduler = make_lr_scheduler(optimizer, total_steps, config.warmup_ratio)

    # Mixed precision setup
    scaler = torch.amp.GradScaler("cuda", enabled=config.use_mixed_precision)
    autocast_ctx_factory = lambda: torch.amp.autocast(
        device_type="cuda", enabled=config.use_mixed_precision
    )

    loss_fn = make_loss_fn(config.class_weights, device)

    os.makedirs(config.checkpoint_dir, exist_ok=True)
    checkpoint_path = os.path.join(config.checkpoint_dir, config.checkpoint_name)

    history = {"train_loss": [], "dev_eer": [], "dev_auc": [], "dev_accuracy": []}
    best_eer = float("inf")
    epochs_without_improvement = 0
    global_step = 0

    for epoch in range(config.epochs):
        model.train()
        epoch_start = time.time()
        epoch_losses = []

        pbar = tqdm(train_loader, desc=f"Epoch {epoch+1}/{config.epochs}")
        for waveforms, labels, utt_ids in pbar:
            waveforms = waveforms.to(device, non_blocking=True)
            labels = labels.to(device, non_blocking=True)

            optimizer.zero_grad()
            with autocast_ctx_factory():
                logits = model(waveforms)
                loss = loss_fn(logits, labels)

            if config.use_mixed_precision:
                scaler.scale(loss).backward()
                scaler.unscale_(optimizer)
                torch.nn.utils.clip_grad_norm_(trainable, config.grad_clip)
                scaler.step(optimizer)
                scaler.update()
            else:
                loss.backward()
                torch.nn.utils.clip_grad_norm_(trainable, config.grad_clip)
                optimizer.step()

            if scheduler is not None:
                scheduler.step()

            epoch_losses.append(loss.item())
            global_step += 1

            if wandb is not None and global_step % config.log_every_n_steps == 0:
                log_data = {
                    "train/step_loss": loss.item(),
                    "train/global_step": global_step,
                    "train/lr": optimizer.param_groups[0]["lr"],
                }
                wandb.log(log_data)

            pbar.set_postfix(loss=f"{loss.item():.4f}",
                            lr=f"{optimizer.param_groups[0]['lr']:.2e}")

        train_loss = float(np.mean(epoch_losses))
        dev_metrics = evaluate(
            model, dev_loader, device,
            desc=f"Epoch {epoch+1} dev",
            use_mixed_precision=config.use_mixed_precision,
        )
        epoch_time = time.time() - epoch_start

        history["train_loss"].append(train_loss)
        history["dev_eer"].append(dev_metrics["eer"])
        history["dev_auc"].append(dev_metrics["auc"])
        history["dev_accuracy"].append(dev_metrics["accuracy"])

        print(f"\\nEpoch {epoch+1}/{config.epochs} ({epoch_time:.0f}s)")
        print(f"  train_loss:   {train_loss:.4f}")
        print(f"  dev_eer:      {dev_metrics['eer']*100:.2f}%")
        print(f"  dev_auc:      {dev_metrics['auc']:.4f}")
        print(f"  dev_accuracy: {dev_metrics['accuracy']*100:.2f}%")

        if wandb is not None:
            wandb.log({
                "epoch": epoch + 1,
                "train/epoch_loss": train_loss,
                "dev/eer": dev_metrics["eer"],
                "dev/auc": dev_metrics["auc"],
                "dev/accuracy": dev_metrics["accuracy"],
                "dev/loss": dev_metrics["loss"],
            })

        if dev_metrics["eer"] < best_eer:
            best_eer = dev_metrics["eer"]
            epochs_without_improvement = 0
            torch.save({
                "epoch": epoch + 1,
                "model_state_dict": model.state_dict(),
                "optimizer_state_dict": optimizer.state_dict(),
                "scheduler_state_dict": scheduler.state_dict() if scheduler else None,
                "best_eer": best_eer,
                "config": vars(config),
            }, checkpoint_path)
            print(f"  → Saved best checkpoint (EER={best_eer*100:.2f}%)")
        else:
            epochs_without_improvement += 1
            print(f"  No improvement for {epochs_without_improvement} epoch(s)")

        if epochs_without_improvement >= config.early_stopping_patience:
            print(f"\\nEarly stopping after {epoch+1} epochs.")
            break

    if wandb is not None:
        wandb.summary["best_dev_eer"] = best_eer
        wandb.finish()

    return {"history": history, "best_eer": best_eer, "checkpoint_path": checkpoint_path}
'''

PATH = '/content/deepfake-audio-detection/src/training/trainer.py'
with open(PATH, 'w') as f:
    f.write(TRAINER_PY)
print(f"Wrote {PATH} ({len(TRAINER_PY)} bytes)")

Wrote /content/deepfake-audio-detection/src/training/trainer.py (9809 bytes)


In [8]:
import sys, importlib

# Reload the updated trainer
for mod in ['src.training.trainer']:
    if mod in sys.modules:
        importlib.reload(sys.modules[mod])
from src.training.trainer import TrainConfig, train

# Build tiny class-balanced subsets (lessons from yesterday)
import torchaudio
from tqdm import tqdm
from torch.utils.data import DataLoader

print("Building tiny class-balanced subsets for smoke test...\n")

train_bonafide = [u for u in splits['train'] if u.label == 'bonafide'][:50]
train_spoof    = [u for u in splits['train'] if u.label == 'spoof'][:50]
train_subset   = train_bonafide + train_spoof

dev_bonafide = [u for u in splits['dev'] if u.label == 'bonafide'][:50]
dev_spoof    = [u for u in splits['dev'] if u.label == 'spoof'][:50]
dev_subset   = dev_bonafide + dev_spoof

# Measure durations
def measure_durs(utts, name):
    durs = []
    for u in tqdm(utts, desc=name):
        w, _ = torchaudio.load(u.flac_path)
        durs.append(w.shape[1])
    return durs

train_durs = measure_durs(train_subset, "Train durs")
dev_durs   = measure_durs(dev_subset,   "Dev durs")

train_ds = ASVspoofDataset(train_subset, durations_samples=train_durs)
dev_ds   = ASVspoofDataset(dev_subset,   durations_samples=dev_durs)

print(f"\nTrain: {len(train_ds)} windows, Dev: {len(dev_ds)} windows")

train_loader = DataLoader(train_ds, batch_size=8, shuffle=True, num_workers=2, pin_memory=True)
dev_loader   = DataLoader(dev_ds,   batch_size=8, shuffle=False, num_workers=2, pin_memory=True)

# Smoke-test config: mixed precision ON, scheduler ON, 1 epoch
config = TrainConfig(
    learning_rate=1e-5,        # Stage 2 LR
    batch_size=8,
    epochs=1,
    weight_decay=0.01,
    grad_clip=1.0,
    early_stopping_patience=99,
    checkpoint_dir='/content/drive/MyDrive/deepfake_audio/checkpoints',
    checkpoint_name='stage2_smoke.pt',
    log_every_n_steps=2,
    use_wandb=True,
    wandb_run_name='stage2-smoke',
    class_weights=None,
    use_mixed_precision=True,
    use_lr_scheduler=True,
    warmup_ratio=0.1,
)

print(f"\nRunning Stage 2 smoke test (1 epoch, batch 8, mixed precision ON)...")
print("Should take 30-60 seconds.\n")

# Model is already loaded with Stage 1 weights and unfrozen top 2 layers
result = train(model, train_loader, dev_loader, config, device='cuda')

print(f"\nSmoke test complete.")
print(f"Best EER: {result['best_eer']*100:.2f}%")

Building tiny class-balanced subsets for smoke test...



Dev durs: 100%|██████████| 100/100 [00:00<00:00, 196.22it/s]



Train: 122 windows, Dev: 133 windows

Running Stage 2 smoke test (1 epoch, batch 8, mixed precision ON)...
Should take 30-60 seconds.



/usr/local/lib/python3.12/dist-packages/notebook/notebookapp.py:191: SyntaxWarning: invalid escape sequence '\/'
  | |_| | '_ \/ _` / _` |  _/ -_)
wandb: [wandb.login()] Loaded credentials for https://api.wandb.ai from WANDB_API_KEY.
wandb: Currently logged in as: sara-jaffrani17 (sara-jaffrani17-dlp) to https://api.wandb.ai. Use `wandb login --relogin` to force relogin


Epoch 1/1: 100%|██████████| 16/16 [00:05<00:00,  2.75it/s, loss=0.0026, lr=0.00e+00]



Epoch 1/1 (7s)
  train_loss:   0.1290
  dev_eer:      2.00%
  dev_auc:      0.9952
  dev_accuracy: 97.00%
  → Saved best checkpoint (EER=2.00%)


dev/accuracy,▁
dev/auc,▁
dev/eer,▁
dev/loss,▁
epoch,▁
train/epoch_loss,▁
train/global_step,▁▂▃▄▅▆▇█
train/lr,█▇▆▅▄▃▂▁
train/step_loss,▃▆█▄▄▁▅▁
best_dev_eer,0.02
dev/accuracy,0.97



Smoke test complete.
Best EER: 2.00%


Step 5: Plan the full Stage 2 training run
Before kicking off the long run, let me lay out exactly what we're configuring and why. This is your interview answer pre-loaded.
Hyperparameters:

Learning rate: 1e-5 (100x lower than Stage 1's 1e-3) — fine-tuning pretrained weights requires gentle adjustments
Warmup: 10% of total steps — first ~1,000 steps ramp LR from 0 to 1e-5, then linear decay
Mixed precision: ON — ~30-40% speedup on T4
Batch size: 16 (down from 32) — mixed precision lets us keep this without OOM, and smaller batches give more gradient noise which helps fine-tuning
Epochs: 10 with early stopping (patience 3)
Weight decay: 0.01
Class weights: same as Stage 1 (bonafide=4.92, spoof=0.56)
Gradient clipping: 1.0

Starting point: Stage 1 checkpoint weights are already loaded into model. Top 2 layers are already unfrozen (14M trainable params).
Expected wall clock: ~45-60 min/epoch on T4 with mixed precision. 10 epochs ≈ 7.5-10 hours total. Early stopping might cut it to 5-7 hours if convergence is fast.
Expected dev EER: Stage 1 was 10.09%. Stage 2 target is 3-7%. Anything below 5% is excellent. If Stage 2 doesn't beat Stage 1 by ≥2 percentage points, we'd debug rather than tune.

In [9]:
import sys, importlib, time
import torch
import torchaudio
from tqdm import tqdm
from torch.utils.data import DataLoader

# Reload modules in case anything was edited
for mod in ['src.data.protocols', 'src.data.preprocessing', 'src.data.dataset',
            'src.models.wav2vec_classifier', 'src.evaluation.metrics',
            'src.training.trainer']:
    if mod in sys.modules:
        importlib.reload(sys.modules[mod])

from src.training.trainer import TrainConfig, train

# ----- Step A: Build full datasets and loaders -----
print("Measuring durations on full train set...")
train_durs = []
for u in tqdm(splits['train'], desc="Train"):
    w, _ = torchaudio.load(u.flac_path)
    train_durs.append(w.shape[1])

print("Measuring durations on full dev set...")
dev_durs = []
for u in tqdm(splits['dev'], desc="Dev"):
    w, _ = torchaudio.load(u.flac_path)
    dev_durs.append(w.shape[1])

train_ds = ASVspoofDataset(splits['train'], durations_samples=train_durs)
dev_ds   = ASVspoofDataset(splits['dev'],   durations_samples=dev_durs)

print(f"\nTrain: {len(train_ds):,} windows from {len(splits['train']):,} utterances")
print(f"Dev:   {len(dev_ds):,} windows from {len(splits['dev']):,} utterances")

# Stage 2 batch size: 16 (smaller than Stage 1's 32)
# Why: fine-tuning benefits from noisier gradients; also leaves headroom for unfrozen-layer activations
train_loader = DataLoader(train_ds, batch_size=16, shuffle=True,  num_workers=2, pin_memory=True)
dev_loader   = DataLoader(dev_ds,   batch_size=16, shuffle=False, num_workers=2, pin_memory=True)

# ----- Step B: Class weights (same as Stage 1) -----
n_bonafide = sum(1 for u in splits['train'] if u.label == 'bonafide')
n_spoof    = len(splits['train']) - n_bonafide
total = n_bonafide + n_spoof
w_bonafide = total / (2 * n_bonafide)
w_spoof    = total / (2 * n_spoof)
class_weights = [w_bonafide, w_spoof]
print(f"\nClass weights: bonafide={w_bonafide:.3f}, spoof={w_spoof:.3f}")

# ----- Step C: Confirm model state -----
trainable = sum(p.numel() for p in model.parameters() if p.requires_grad)
print(f"\nModel state:")
print(f"  Total params:     {sum(p.numel() for p in model.parameters()):,}")
print(f"  Trainable params: {trainable:,}  ({100*trainable/94373250:.2f}% of total)")
print(f"  Stage 1 weights:  loaded (best dev EER 10.09%)")

# ----- Step D: Stage 2 config -----
config = TrainConfig(
    learning_rate=1e-5,
    batch_size=16,
    epochs=10,
    weight_decay=0.01,
    grad_clip=1.0,
    early_stopping_patience=3,
    checkpoint_dir='/content/drive/MyDrive/deepfake_audio/checkpoints',
    checkpoint_name='stage2_best.pt',
    log_every_n_steps=20,
    use_wandb=True,
    wandb_run_name='stage2-full',
    class_weights=class_weights,
    use_mixed_precision=True,
    use_lr_scheduler=True,
    warmup_ratio=0.1,
)

print(f"\n{'='*70}")
print(f"  STAGE 2 FULL TRAINING — 10 epochs, top 2 layers unfrozen")
print(f"  LR: 1e-5 with 10% warmup + linear decay")
print(f"  Mixed precision: ON")
print(f"  Early stopping: patience 3")
print(f"  Starting from Stage 1 checkpoint (10.09% EER)")
print(f"{'='*70}\n")

start = time.time()
result = train(model, train_loader, dev_loader, config, device='cuda')
elapsed_hr = (time.time() - start) / 3600

print(f"\n{'='*70}")
print(f"  TRAINING COMPLETE in {elapsed_hr:.2f} hours")
print(f"{'='*70}")
print(f"Best dev EER: {result['best_eer']*100:.2f}%")
print(f"Stage 1 EER:  10.09%")
print(f"Improvement:  {(0.1009 - result['best_eer'])*100:.2f} percentage points")
print(f"Checkpoint:   {result['checkpoint_path']}")

print(f"\nEpoch-by-epoch:")
print(f"{'Epoch':<6} {'Train Loss':<12} {'Dev EER':<10} {'Dev AUC':<10} {'Dev Acc':<10}")
print("-" * 50)
hist = result['history']
for i, (loss, eer, auc, acc) in enumerate(zip(
    hist['train_loss'], hist['dev_eer'], hist['dev_auc'], hist['dev_accuracy']
)):
    print(f"{i+1:<6} {loss:<12.4f} {eer*100:<10.2f}% {auc:<10.4f} {acc*100:<10.2f}%")

Measuring durations on full train set...


Train: 100%|██████████| 25380/25380 [01:21<00:00, 312.15it/s]


Measuring durations on full dev set...


Dev: 100%|██████████| 24844/24844 [01:19<00:00, 312.43it/s]



Train: 34,036 windows from 25,380 utterances
Dev:   34,047 windows from 24,844 utterances

Class weights: bonafide=4.919, spoof=0.557

Model state:
  Total params:     94,373,250
  Trainable params: 14,178,818  (15.02% of total)
  Stage 1 weights:  loaded (best dev EER 10.09%)

  STAGE 2 FULL TRAINING — 10 epochs, top 2 layers unfrozen
  LR: 1e-5 with 10% warmup + linear decay
  Mixed precision: ON
  Early stopping: patience 3
  Starting from Stage 1 checkpoint (10.09% EER)



Epoch 1/10: 100%|██████████| 2128/2128 [13:44<00:00,  2.58it/s, loss=0.0000, lr=1.00e-05]



Epoch 1/10 (1094s)
  train_loss:   0.1501
  dev_eer:      3.29%
  dev_auc:      0.9965
  dev_accuracy: 96.72%
  → Saved best checkpoint (EER=3.29%)


Epoch 2/10: 100%|██████████| 2128/2128 [13:05<00:00,  2.71it/s, loss=0.0000, lr=8.89e-06]



Epoch 2/10 (1054s)
  train_loss:   0.0734
  dev_eer:      1.45%
  dev_auc:      0.9987
  dev_accuracy: 98.55%
  → Saved best checkpoint (EER=1.45%)


Epoch 3/10: 100%|██████████| 2128/2128 [12:58<00:00,  2.73it/s, loss=0.0000, lr=7.78e-06]



Epoch 3/10 (1047s)
  train_loss:   0.0504
  dev_eer:      1.22%
  dev_auc:      0.9990
  dev_accuracy: 98.79%
  → Saved best checkpoint (EER=1.22%)


Epoch 4/10: 100%|██████████| 2128/2128 [12:57<00:00,  2.74it/s, loss=0.0000, lr=6.67e-06]



Epoch 4/10 (1046s)
  train_loss:   0.0506
  dev_eer:      1.42%
  dev_auc:      0.9984
  dev_accuracy: 98.61%
  No improvement for 1 epoch(s)


Epoch 5/10: 100%|██████████| 2128/2128 [12:58<00:00,  2.73it/s, loss=0.0000, lr=5.56e-06]



Epoch 5/10 (1047s)
  train_loss:   0.0405
  dev_eer:      1.19%
  dev_auc:      0.9987
  dev_accuracy: 98.80%
  → Saved best checkpoint (EER=1.19%)


Epoch 6/10: 100%|██████████| 2128/2128 [12:56<00:00,  2.74it/s, loss=0.0000, lr=4.44e-06]



Epoch 6/10 (1045s)
  train_loss:   0.0191
  dev_eer:      1.02%
  dev_auc:      0.9985
  dev_accuracy: 98.97%
  → Saved best checkpoint (EER=1.02%)


Epoch 7/10: 100%|██████████| 2128/2128 [12:56<00:00,  2.74it/s, loss=0.0000, lr=3.33e-06]



Epoch 7/10 (1045s)
  train_loss:   0.0195
  dev_eer:      0.84%
  dev_auc:      0.9986
  dev_accuracy: 99.17%
  → Saved best checkpoint (EER=0.84%)


Epoch 8/10: 100%|██████████| 2128/2128 [12:57<00:00,  2.74it/s, loss=0.0000, lr=2.22e-06]



Epoch 8/10 (1046s)
  train_loss:   0.0097
  dev_eer:      0.79%
  dev_auc:      0.9984
  dev_accuracy: 99.21%
  → Saved best checkpoint (EER=0.79%)


Epoch 9/10: 100%|██████████| 2128/2128 [12:55<00:00,  2.74it/s, loss=0.0000, lr=1.11e-06]



Epoch 9/10 (1043s)
  train_loss:   0.0113
  dev_eer:      0.69%
  dev_auc:      0.9984
  dev_accuracy: 99.31%
  → Saved best checkpoint (EER=0.69%)


Epoch 10/10: 100%|██████████| 2128/2128 [12:54<00:00,  2.75it/s, loss=0.0000, lr=0.00e+00]



Epoch 10/10 (1044s)
  train_loss:   0.0069
  dev_eer:      0.82%
  dev_auc:      0.9988
  dev_accuracy: 99.15%
  No improvement for 1 epoch(s)


dev/accuracy,▁▆▇▆▇▇████
dev/auc,▁▇█▆▇▇▇▆▆▇
dev/eer,█▃▂▃▂▂▁▁▁▁
dev/loss,█▂▂▄▃▃▁▁▁▁
epoch,▁▂▃▃▄▅▆▆▇█
train/epoch_loss,█▄▃▃▃▂▂▁▁▁
train/global_step,▁▁▁▂▂▂▃▃▃▃▃▃▃▃▃▄▄▄▄▄▄▅▅▅▅▅▅▅▆▆▆▆▆▇▇█████
train/lr,▂▆███▇▇▇▇▆▆▅▅▅▅▅▅▅▅▅▄▄▄▄▄▄▃▃▃▃▃▂▂▂▂▂▁▁▁▁
train/step_loss,▅▄▁█▁▁▁▁▁█▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁
best_dev_eer,0.00694
dev/accuracy,0.99147



  TRAINING COMPLETE in 2.93 hours
Best dev EER: 0.69%
Stage 1 EER:  10.09%
Improvement:  9.40 percentage points
Checkpoint:   /content/drive/MyDrive/deepfake_audio/checkpoints/stage2_best.pt

Epoch-by-epoch:
Epoch  Train Loss   Dev EER    Dev AUC    Dev Acc   
--------------------------------------------------
1      0.1501       3.29      % 0.9965     96.72     %
2      0.0734       1.45      % 0.9987     98.55     %
3      0.0504       1.22      % 0.9990     98.79     %
4      0.0506       1.42      % 0.9984     98.61     %
5      0.0405       1.19      % 0.9987     98.80     %
6      0.0191       1.02      % 0.9985     98.97     %
7      0.0195       0.84      % 0.9986     99.17     %
8      0.0097       0.79      % 0.9984     99.21     %
9      0.0113       0.69      % 0.9984     99.31     %
10     0.0069       0.82      % 0.9988     99.15     %


Things that are normal:

Loss starts higher than expected (Stage 1 weights are tuned for the head's loss, not the unfrozen-layer loss)
LR ramps up during warmup (visible in tqdm lr= field)
First epoch takes 45-60 min
Loss is choppy — Stage 2 batches are smaller, more noise

Things that are NOT normal — interrupt and ping me:

Loss is NaN or Inf
Loss is increasing across many batches (>200 batches of monotonic increase)
Out-of-memory error
The cell crashes silently

The critical numbers:

After epoch 1, paste the dev metrics line. I'm looking for dev_eer in the 5-12% range. Inside that range = healthy. Outside = we investigate.


EpochTrain LossDev EERNote10.1503.29%Massive jump from Stage 1's 10.09%20.0731.45%Still rapid improvement30.0501.22%Slowing down40.0511.42%First non-improvement50.0401.19%Recovered60.0191.02%Crossed 1%70.0200.84%Below 1% threshold80.0100.79%Still improving90.0110.69%Best100.0070.82%Slight overfitting starts
Three things worth noticing:

Massive Stage 1 → Stage 2 leap. 10.09% → 3.29% in one epoch. Confirms that pretrained features are valuable but task-specific fine-tuning matters a lot.
Train loss kept dropping but dev EER plateaued around 0.7-0.8%. The model is starting to overfit at epochs 9-10. Training longer would not have helped.
The LR schedule worked exactly as designed. Warmup → peak → decay → near zero, with the model converging to its best as the LR cooled down. This is a textbook trajectory.

In [10]:
import torch, os
CKPT = '/content/drive/MyDrive/deepfake_audio/checkpoints/stage2_best.pt'
print(f"Path: {CKPT}")
print(f"Exists: {os.path.exists(CKPT)}")
print(f"Size: {os.path.getsize(CKPT) / 1e6:.1f} MB")

ckpt = torch.load(CKPT, map_location='cpu', weights_only=False)
print(f"\nCheckpoint contents:")
print(f"  Epoch saved:  {ckpt['epoch']}")
print(f"  Best dev EER: {ckpt['best_eer']*100:.4f}%")
print(f"  State dict tensors: {len(ckpt['model_state_dict'])}")
print(f"  Has scheduler state: {'scheduler_state_dict' in ckpt and ckpt['scheduler_state_dict'] is not None}")

Path: /content/drive/MyDrive/deepfake_audio/checkpoints/stage2_best.pt
Exists: True
Size: 491.0 MB

Checkpoint contents:
  Epoch saved:  9
  Best dev EER: 0.6941%
  State dict tensors: 213
  Has scheduler state: True


In [11]:
import json, os
from datetime import datetime

RESULTS_DIR = '/content/deepfake-audio-detection/results/metrics'
os.makedirs(RESULTS_DIR, exist_ok=True)

stage2_results = {
    "phase": "Stage 2 — fine-tuning top 2 transformer layers",
    "completed_at": datetime.now().isoformat(),
    "model": "facebook/wav2vec2-base",
    "starting_checkpoint": "stage1_best.pt (dev EER 10.09%)",
    "trainable_params": 14178818,
    "total_params": 94373250,
    "trainable_percent": 15.02,
    "training": {
        "epochs": 10,
        "batch_size": 16,
        "learning_rate": 1e-5,
        "warmup_ratio": 0.1,
        "lr_schedule": "linear warmup + linear decay to zero",
        "weight_decay": 0.01,
        "grad_clip": 1.0,
        "class_weights": {"bonafide": 4.919, "spoof": 0.557},
        "loss": "weighted_cross_entropy",
        "mixed_precision": True,
        "wall_clock_hours": 2.93,
        "early_stopping_patience": 3,
        "early_stopping_triggered": False,
    },
    "data": {
        "train_utterances": 25380,
        "train_windows": 34036,
        "dev_utterances": 24844,
        "dev_windows": 34047,
    },
    "epoch_history": [
        {"epoch": 1,  "train_loss": 0.1501, "dev_eer": 0.0329, "dev_auc": 0.9965, "dev_accuracy": 0.9672},
        {"epoch": 2,  "train_loss": 0.0734, "dev_eer": 0.0145, "dev_auc": 0.9987, "dev_accuracy": 0.9855},
        {"epoch": 3,  "train_loss": 0.0504, "dev_eer": 0.0122, "dev_auc": 0.9990, "dev_accuracy": 0.9879},
        {"epoch": 4,  "train_loss": 0.0506, "dev_eer": 0.0142, "dev_auc": 0.9984, "dev_accuracy": 0.9861},
        {"epoch": 5,  "train_loss": 0.0405, "dev_eer": 0.0119, "dev_auc": 0.9987, "dev_accuracy": 0.9880},
        {"epoch": 6,  "train_loss": 0.0191, "dev_eer": 0.0102, "dev_auc": 0.9985, "dev_accuracy": 0.9897},
        {"epoch": 7,  "train_loss": 0.0195, "dev_eer": 0.0084, "dev_auc": 0.9986, "dev_accuracy": 0.9917},
        {"epoch": 8,  "train_loss": 0.0097, "dev_eer": 0.0079, "dev_auc": 0.9984, "dev_accuracy": 0.9921},
        {"epoch": 9,  "train_loss": 0.0113, "dev_eer": 0.0069, "dev_auc": 0.9984, "dev_accuracy": 0.9931},
        {"epoch": 10, "train_loss": 0.0069, "dev_eer": 0.0082, "dev_auc": 0.9988, "dev_accuracy": 0.9915},
    ],
    "best": {
        "dev_eer": 0.0069,
        "dev_auc": 0.9984,
        "dev_accuracy": 0.9931,
        "epoch": 9,
        "checkpoint": "/content/drive/MyDrive/deepfake_audio/checkpoints/stage2_best.pt",
    },
    "improvement_over_stage1": {
        "stage1_dev_eer": 0.1009,
        "stage2_dev_eer": 0.0069,
        "absolute_improvement_percentage_points": 9.40,
        "relative_error_reduction_percent": 93.16,
    },
    "wandb_run": "https://wandb.ai/sara-jaffrani17-dlp/deepfake-audio-detection/runs/l1q4dvsx",
    "notes": [
        "Stage 2 unfroze the top 2 transformer encoder layers (10, 11) plus final layer norm.",
        "14.18M trainable parameters (15.02% of total model).",
        "Started from Stage 1 checkpoint with 10.09% dev EER.",
        "Best dev EER 0.69% achieved at epoch 9, with mild overfitting beginning at epoch 10.",
        "Training stable throughout: no NaN, no loss spikes, smooth convergence.",
        "Dev accuracy 99.31%, dev AUC 0.9984 — model is now in the territory of strong published systems on ASVspoof 2019 LA.",
        "IMPORTANT CAVEAT: Dev set contains same attacks as training (A01-A06).",
        "Eval set contains unseen attacks (A07-A19). True generalization test is forthcoming in Phase 5.",
    ]
}

OUTPUT = f"{RESULTS_DIR}/stage2_results.json"
with open(OUTPUT, 'w') as f:
    json.dump(stage2_results, f, indent=2)

print(f"Wrote {OUTPUT}")
print(f"Size: {os.path.getsize(OUTPUT)} bytes")

Wrote /content/deepfake-audio-detection/results/metrics/stage2_results.json
Size: 3492 bytes


In [13]:
from google.colab import userdata
import os

GITHUB_TOKEN = userdata.get('GITHUB_TOKEN')
os.chdir('/content/deepfake-audio-detection')

# Configure git identity (in case it got reset after re-clone)
!git config user.email "95262824+Saracasm@users.noreply.github.com"
!git config user.name "Sara Iqbal"

# Stage all the changes
!git add results/metrics/stage2_results.json
!git add src/training/trainer.py  # the updated version with mixed precision + scheduler
!git status

# Commit
!git commit -m "Phase 4: Stage 2 fine-tuning — 0.69% dev EER (93% error reduction vs Stage 1)"

# Push
push_url = f"https://Saracasm:{GITHUB_TOKEN}@github.com/Saracasm/deepfake-audio-detection.git"
!git push {push_url} main 2>&1 | grep -v {GITHUB_TOKEN}

On branch main
Your branch is up to date with 'origin/main'.

Changes to be committed:
  (use "git restore --staged <file>..." to unstage)
	new file:   results/metrics/stage2_results.json
	modified:   src/training/trainer.py

[main 642a1a8] Phase 4: Stage 2 fine-tuning — 0.69% dev EER (93% error reduction vs Stage 1)
 2 files changed, 208 insertions(+), 53 deletions(-)
 create mode 100644 results/metrics/stage2_results.json
To https://github.com/Saracasm/deepfake-audio-detection.git
   5e3cbcc..642a1a8  main -> main


In [14]:
import sys, importlib, time
import torch
import torchaudio
import numpy as np
from tqdm import tqdm
from torch.utils.data import DataLoader
from collections import defaultdict

# Reload modules
for mod in ['src.data.protocols', 'src.data.preprocessing', 'src.data.dataset',
            'src.models.wav2vec_classifier', 'src.evaluation.metrics']:
    if mod in sys.modules:
        importlib.reload(sys.modules[mod])

from src.evaluation.metrics import compute_eer, compute_auc, aggregate_window_scores_to_utterance

# ----- Step A: Measure eval durations -----
print(f"Eval set: {len(splits['eval']):,} utterances")
print("Measuring durations on full eval set...")

eval_durs = []
for u in tqdm(splits['eval'], desc="Eval"):
    w, _ = torchaudio.load(u.flac_path)
    eval_durs.append(w.shape[1])

eval_ds = ASVspoofDataset(splits['eval'], durations_samples=eval_durs)
eval_loader = DataLoader(eval_ds, batch_size=16, shuffle=False, num_workers=2, pin_memory=True)

print(f"\nEval dataset: {len(eval_ds):,} windows from {len(splits['eval']):,} utterances")
inflation = len(eval_ds) / len(splits['eval'])
print(f"Window inflation factor: {inflation:.2f}x")

# ----- Step B: Make sure we're using the Stage 2 model state -----
# (The model variable in memory should already have Stage 2 weights from training,
#  but let's reload from checkpoint to be 100% sure we're evaluating the BEST checkpoint
#  not the LAST checkpoint)
print("\nLoading Stage 2 best checkpoint into model...")
CKPT_PATH = '/content/drive/MyDrive/deepfake_audio/checkpoints/stage2_best.pt'
ckpt = torch.load(CKPT_PATH, map_location='cuda', weights_only=False)
model.load_state_dict(ckpt['model_state_dict'])
print(f"Loaded checkpoint from epoch {ckpt['epoch']} (best dev EER {ckpt['best_eer']*100:.4f}%)")

# ----- Step C: Run inference on full eval set -----
print(f"\nRunning inference (mixed precision, batch=16)...")
print(f"Estimated time: 25-40 min on T4\n")

model.eval()
all_window_scores = []
all_window_labels = []
all_window_utts = []
all_window_attacks = []  # extra: track attack ID per window for per-attack analysis

# Build a lookup from utterance_id to attack_id
utt_to_attack = {u.utterance_id: u.attack_id for u in splits['eval']}

start = time.time()
with torch.no_grad():
    autocast_ctx = torch.amp.autocast(device_type='cuda', enabled=True)
    for waveforms, labels, utt_ids in tqdm(eval_loader, desc="Eval inference"):
        waveforms = waveforms.to('cuda', non_blocking=True)
        with autocast_ctx:
            logits = model(waveforms)
        probs = torch.softmax(logits.float(), dim=-1)
        spoof_probs = probs[:, 1].detach().cpu().numpy()

        all_window_scores.extend(spoof_probs.tolist())
        all_window_labels.extend(labels.tolist())
        all_window_utts.extend(list(utt_ids))
        all_window_attacks.extend([utt_to_attack[u] for u in utt_ids])

inference_minutes = (time.time() - start) / 60
print(f"\nInference complete in {inference_minutes:.1f} min over {len(all_window_scores):,} windows.")

# ----- Step D: Aggregate per-window to per-utterance -----
print("\nAggregating window scores to utterance scores (mean)...")
utt_scores, utt_ids_sorted = aggregate_window_scores_to_utterance(
    np.array(all_window_scores), all_window_utts, method="mean",
)

# Get per-utterance labels and attacks (use first occurrence of each utt)
utt_label_map = {}
utt_attack_map = {}
for s, l, u, a in zip(all_window_scores, all_window_labels, all_window_utts, all_window_attacks):
    if u not in utt_label_map:
        utt_label_map[u] = l
        utt_attack_map[u] = a
utt_labels = np.array([utt_label_map[u] for u in utt_ids_sorted])
utt_attacks = np.array([utt_attack_map[u] for u in utt_ids_sorted])

# ----- Step E: Overall metrics -----
print(f"\n{'='*70}")
print(f"  PRIMARY EVALUATION — ASVspoof 2019 LA Eval Set")
print(f"{'='*70}")
print(f"Utterances: {len(utt_scores):,}")
n_bona = int((utt_labels == 0).sum())
n_spoof = int((utt_labels == 1).sum())
print(f"Bonafide:   {n_bona:,}")
print(f"Spoof:      {n_spoof:,}")

eer, threshold = compute_eer(utt_scores, utt_labels)
auc = compute_auc(utt_scores, utt_labels)
preds = (utt_scores > threshold).astype(int)
acc = float((preds == utt_labels).mean())

print(f"\nOverall results (Stage 2 model):")
print(f"  EER:       {eer*100:.4f}%")
print(f"  AUC:       {auc:.4f}")
print(f"  Accuracy:  {acc*100:.2f}%")
print(f"  Threshold: {threshold:.4f}")

# Comparison
print(f"\nComparison:")
print(f"  Stage 2 dev EER (seen attacks A01-A06):    0.69%")
print(f"  Stage 2 eval EER (unseen attacks A07-A19): {eer*100:.2f}%")
gen_gap = (eer - 0.0069) * 100
print(f"  Generalization gap:                         {gen_gap:.2f} percentage points")

# ----- Step F: Per-attack EER for the eval set -----
print(f"\n{'='*70}")
print(f"  PER-ATTACK EER BREAKDOWN (eval set, A07-A19)")
print(f"{'='*70}")

# For per-attack EER, we treat each attack vs all bonafide
bonafide_scores = utt_scores[utt_labels == 0]

per_attack_results = {}
attack_ids_in_eval = sorted(set(a for a in utt_attacks if a != '-'))

for attack in attack_ids_in_eval:
    mask = (utt_attacks == attack)
    attack_scores = utt_scores[mask]
    n = len(attack_scores)
    # Compute EER between this attack and all bonafide
    combined_scores = np.concatenate([bonafide_scores, attack_scores])
    combined_labels = np.concatenate([
        np.zeros(len(bonafide_scores)),  # bonafide = 0
        np.ones(n),                       # spoof = 1
    ])
    a_eer, _ = compute_eer(combined_scores, combined_labels)
    per_attack_results[attack] = {"n": int(n), "eer": float(a_eer)}
    print(f"  {attack}: n={n:>5,}  EER={a_eer*100:>7.2f}%")

# Summary stats
attack_eers = [v["eer"] for v in per_attack_results.values()]
print(f"\nAcross {len(attack_eers)} unseen attacks:")
print(f"  Mean EER:   {np.mean(attack_eers)*100:.2f}%")
print(f"  Median EER: {np.median(attack_eers)*100:.2f}%")
print(f"  Worst (highest EER): {np.max(attack_eers)*100:.2f}%")
print(f"  Best (lowest EER):   {np.min(attack_eers)*100:.2f}%")

# Save the raw scores so we can re-analyze without re-inferring
import os
os.makedirs('/content/deepfake-audio-detection/results/scores', exist_ok=True)
SCORES_PATH = '/content/deepfake-audio-detection/results/scores/stage2_eval2019.npz'
np.savez(SCORES_PATH,
    utt_ids=np.array(utt_ids_sorted),
    utt_scores=utt_scores,
    utt_labels=utt_labels,
    utt_attacks=utt_attacks,
)
print(f"\nRaw scores saved to {SCORES_PATH}")
print(f"(Use this to recompute metrics later without re-inferring)")

Eval set: 71,237 utterances
Measuring durations on full eval set...


Eval: 100%|██████████| 71237/71237 [04:02<00:00, 294.00it/s]



Eval dataset: 91,386 windows from 71,237 utterances
Window inflation factor: 1.28x

Loading Stage 2 best checkpoint into model...
Loaded checkpoint from epoch 9 (best dev EER 0.6941%)

Running inference (mixed precision, batch=16)...
Estimated time: 25-40 min on T4



Eval inference: 100%|██████████| 5712/5712 [11:59<00:00,  7.94it/s]



Inference complete in 12.0 min over 91,386 windows.

Aggregating window scores to utterance scores (mean)...

  PRIMARY EVALUATION — ASVspoof 2019 LA Eval Set
Utterances: 71,237
Bonafide:   7,355
Spoof:      63,882

Overall results (Stage 2 model):
  EER:       5.5491%
  AUC:       0.9885
  Accuracy:  94.45%
  Threshold: 0.0000

Comparison:
  Stage 2 dev EER (seen attacks A01-A06):    0.69%
  Stage 2 eval EER (unseen attacks A07-A19): 5.55%
  Generalization gap:                         4.86 percentage points

  PER-ATTACK EER BREAKDOWN (eval set, A07-A19)
  A07: n=4,914  EER=   5.81%
  A08: n=4,914  EER=   0.63%
  A09: n=4,914  EER=   0.60%
  A10: n=4,914  EER=  15.54%
  A11: n=4,914  EER=   1.05%
  A12: n=4,914  EER=   0.99%
  A13: n=4,914  EER=   0.24%
  A14: n=4,914  EER=   6.05%
  A15: n=4,914  EER=   7.53%
  A16: n=4,914  EER=   2.31%
  A17: n=4,914  EER=   3.82%
  A18: n=4,914  EER=   2.72%
  A19: n=4,914  EER=   3.79%

Across 13 unseen attacks:
  Mean EER:   3.93%
  Median EER:

In [15]:
import json
import os
from datetime import datetime

RESULTS_DIR = '/content/deepfake-audio-detection/results/metrics'

eval_results = {
    "phase": "Phase 5 — Primary Evaluation on ASVspoof 2019 LA Eval Set",
    "completed_at": datetime.now().isoformat(),
    "model_checkpoint": "/content/drive/MyDrive/deepfake_audio/checkpoints/stage2_best.pt",
    "model_dev_eer": 0.0069,
    "evaluation_dataset": {
        "name": "ASVspoof 2019 LA Eval Set",
        "utterances": 71237,
        "windows": 91386,
        "bonafide_count": 7355,
        "spoof_count": 63882,
        "attack_ids_unseen": ["A07", "A08", "A09", "A10", "A11", "A12", "A13", "A14", "A15", "A16", "A17", "A18", "A19"],
    },
    "inference": {
        "batch_size": 16,
        "mixed_precision": True,
        "wall_clock_minutes": 12.0,
        "windows_per_second": 127,
    },
    "overall_results": {
        "eer": 0.0555,
        "auc": 0.9885,
        "accuracy": 0.9445,
        "threshold": 0.0000,
    },
    "generalization_analysis": {
        "dev_eer_seen_attacks_A01_A06": 0.0069,
        "eval_eer_unseen_attacks_A07_A19": 0.0555,
        "gap_percentage_points": 4.86,
        "interpretation": "Healthy generalization gap. Model retains useful spoofing detection signal on unseen attacks.",
    },
    "per_attack_eer": {
        "A07": 0.0581,
        "A08": 0.0063,
        "A09": 0.0060,
        "A10": 0.1554,
        "A11": 0.0105,
        "A12": 0.0099,
        "A13": 0.0024,
        "A14": 0.0605,
        "A15": 0.0753,
        "A16": 0.0231,
        "A17": 0.0382,
        "A18": 0.0272,
        "A19": 0.0379,
    },
    "per_attack_summary": {
        "n_attacks": 13,
        "samples_per_attack": 4914,
        "mean_eer": 0.0393,
        "median_eer": 0.0272,
        "worst_attack": {"id": "A10", "eer": 0.1554},
        "best_attack": {"id": "A13", "eer": 0.0024},
    },
    "comparisons_to_published_baselines": {
        "asvspoof_2019_lfcc_gmm_baseline_eer": 0.0809,
        "improvement_over_baseline_pp": 2.54,
        "interpretation": "Stage 2 model outperforms the official ASVspoof 2019 LFCC-GMM baseline by 2.54 percentage points.",
    },
    "raw_scores_path": "/content/deepfake-audio-detection/results/scores/stage2_eval2019.npz",
    "wandb_run": "https://wandb.ai/sara-jaffrani17-dlp/deepfake-audio-detection/runs/l1q4dvsx",
    "notes": [
        "Primary evaluation on the ASVspoof 2019 LA eval set (71,237 utterances, A07-A19 unseen attacks).",
        "Overall EER 5.55% beats the official LFCC-GMM baseline (8.09%) by 2.54 pp.",
        "Per-attack analysis reveals A10 as the model's weakest point (15.54% EER, ~3x harder than next-worst).",
        "Median per-attack EER of 2.72% indicates strong generalization on most attack families.",
        "Generalization gap of 4.86 pp from dev (0.69%) to eval (5.55%) is healthy and expected.",
        "Phase 5 next steps: cross-dataset evaluation on ASVspoof 2021 LA and WaveFake.",
    ]
}

OUTPUT = f"{RESULTS_DIR}/stage2_eval2019_results.json"
with open(OUTPUT, 'w') as f:
    json.dump(eval_results, f, indent=2)

print(f"Wrote {OUTPUT}")
print(f"Size: {os.path.getsize(OUTPUT)} bytes")

Wrote /content/deepfake-audio-detection/results/metrics/stage2_eval2019_results.json
Size: 2722 bytes


In [16]:
from google.colab import userdata
import os

GITHUB_TOKEN = userdata.get('GITHUB_TOKEN')
os.chdir('/content/deepfake-audio-detection')

!git config user.email "95262824+Saracasm@users.noreply.github.com"
!git config user.name "Sara Iqbal"

# Stage everything new
!git add results/metrics/stage2_eval2019_results.json
!git add results/scores/stage2_eval2019.npz
!git status

!git commit -m "Phase 5: primary 2019 LA eval — 5.55% EER, beats LFCC-GMM baseline by 2.54pp"

push_url = f"https://Saracasm:{GITHUB_TOKEN}@github.com/Saracasm/deepfake-audio-detection.git"
!git push {push_url} main 2>&1 | grep -v {GITHUB_TOKEN}

On branch main
Your branch is ahead of 'origin/main' by 1 commit.
  (use "git push" to publish your local commits)

Changes to be committed:
  (use "git restore --staged <file>..." to unstage)
	new file:   results/metrics/stage2_eval2019_results.json
	new file:   results/scores/stage2_eval2019.npz

[main 888b5b5] Phase 5: primary 2019 LA eval — 5.55% EER, beats LFCC-GMM baseline by 2.54pp
 2 files changed, 90 insertions(+)
 create mode 100644 results/metrics/stage2_eval2019_results.json
 create mode 100644 results/scores/stage2_eval2019.npz
To https://github.com/Saracasm/deepfake-audio-detection.git
   642a1a8..888b5b5  main -> main


In [1]:
import os, shutil

# Find 02_train_stage2.ipynb in Drive
drive_path = '/content/drive/MyDrive/Colab Notebooks/02_train_stage2.ipynb'
repo_path = '/content/deepfake-audio-detection/notebooks/02_train_stage2.ipynb'

if not os.path.exists(drive_path):
    print(f"NOT FOUND at: {drive_path}")
    print("\nSearching Drive for 02_*.ipynb files:")
    for root, dirs, files in os.walk('/content/drive/MyDrive'):
        if 'deepfake_audio' in root and 'data' in root:
            dirs.clear()
            continue
        for f in files:
            if f.endswith('.ipynb') and ('02' in f or 'stage2' in f.lower() or 'train' in f.lower()):
                full = os.path.join(root, f)
                size_kb = os.path.getsize(full) / 1024
                print(f"  [{size_kb:.1f} KB]  {full}")
else:
    os.makedirs(os.path.dirname(repo_path), exist_ok=True)
    shutil.copy(drive_path, repo_path)
    print(f"Copied to: {repo_path}")
    print(f"Size: {os.path.getsize(repo_path) / 1024:.1f} KB")

NOT FOUND at: /content/drive/MyDrive/Colab Notebooks/02_train_stage2.ipynb

Searching Drive for 02_*.ipynb files:
